src/Workspaces/Core/Portable/Serialization/SerializableSourceText.cs

```diff
@@ -49,21 +52,51 @@ private SerializableSourceText(ITemporaryTextStorageWithName? storage, SourceTex
         {
             Debug.Assert(storage is null != text is null);
 
-            Storage = storage;
-            Text = text;
+            _storage = storage;
+            _text = text;
         }
 
-        public ImmutableArray<byte> GetChecksum()
+        public ImmutableArray<byte> GetChecksum(CancellationToken cancellationToken)
         {
-            return Text?.GetChecksum() ?? Storage!.GetChecksum();
+            SourceText? text;
+            ITemporaryTextStorageWithName? storage;
+            using (_gate.DisposableWait(cancellationToken))
+            {
+                text = _text;
+                storage = _storage;
+            }
+
+            return text?.GetChecksum() ?? storage!.GetChecksum();
         }
 
         public async ValueTask<SourceText> GetTextAsync(CancellationToken cancellationToken)
         {
-            if (Text is not null)
-                return Text;
+            using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+            {
+                // if not computed, then compute and swap over to that value.
+                if (_text == null)
+                {
+                    _text = await _storage!.ReadTextAsync(cancellationToken).ConfigureAwait(false);
+                    _storage = null;
+                }
 
-            return await Storage!.ReadTextAsync(cancellationToken).ConfigureAwait(false);
+                return _text;
+            }
+        }
+
+        public SourceText GetText(CancellationToken cancellationToken)
+        {
+            using (_gate.DisposableWait(cancellationToken))
+            {
+                // if not computed, then compute and swap over to that value.
+                if (_text == null)
+                {
+                    _text = _storage!.ReadText(cancellationToken);
+                    _storage = null;
+                }
+
+                return _text;
+            }
         }
```

Comment:
these functions are the meat of the change.  instead of storing eitehr text/storage for the lifetime of this wrapper.  We start potentially with 'storage' and we atomically swap over to 'text' once first asked.